File: src/uk/ac/shef/dcs/jate/v2/algorithm/RIDF.java
package uk.ac.shef.dcs.jate.v2.algorithm;

import uk.ac.shef.dcs.jate.v2.JATEException;
import uk.ac.shef.dcs.jate.v2.feature.AbstractFeature;
import uk.ac.shef.dcs.jate.v2.feature.FrequencyTermBased;
import uk.ac.shef.dcs.jate.v2.model.JATETerm;
import uk.ac.shef.dcs.jate.v2.model.TermInfo;

import java.util.*;
import java.util.logging.Logger;

/**
 * Residual IDF, see
 * Church, K. and Gale, W. 1995a. Inverse Document Frequency (IDF): A Measure of Deviation from Poisson. In Proceedings of the 3rd Workshop on Very Large Corpora. Cambridge, Massachusetts, USA, pp.121-30.
 */
public class RIDF extends Algorithm{
    private static final Logger LOG = Logger.getLogger(RIDF.class.getName());
    @Override
    public List<JATETerm> execute(Set<String> candidates) throws JATEException {
        AbstractFeature feature = features.get(FrequencyTermBased.class.getName());
        validateFeature(feature, FrequencyTermBased.class);
        FrequencyTermBased fFeature = (FrequencyTermBased) feature;

        double totalDocs = (double) fFeature.getTotalDocs();
        boolean collectInfo=termInfoCollector!=null;
        List<JATETerm> result = new ArrayList<>();

        StringBuilder msg = new StringBuilder("Beginning computing RIDF values,");
        msg.append(", total terms=" + candidates.size());
        LOG.info(msg.toString());
        for(String tString: candidates){
            JATETerm term = new JATETerm(tString);
            int ttf = fFeature.getTTF(tString);
            double cf_over_N = (double) ttf / totalDocs;
            double exponential = Math.exp(0 - cf_over_N);
            double nominator = totalDocs * (1 - exponential);
            double denominator = (double) fFeature.getTermFrequencyInDocument(tString).size();

            if (denominator == 0) {
                denominator=1; //this shouldnt occur. a term that is firstly extracted from the corpus must have a source
            }
            double ridf = Math.log(nominator / denominator) / Math.log(2.0);
            term.setScore(ridf);
            if(collectInfo){
                TermInfo termInfo =termInfoCollector.collect(tString);
                term.setTermInfo(termInfo);
            }
            result.add(term);
        }
        Collections.sort(result);
        LOG.info("Complete");
        return result;
    }
}

File: src/jate/algorithms/ridf.py
"""Residual IDF (RIDF) scoring algorithm."""

from __future__ import annotations

import math
from typing import Any

from jate.algorithms.base import Algorithm
from jate.features import TermFrequency
from jate.models import Candidate, Term, TermExtractionResult


class RIDF(Algorithm):
    """Residual IDF.

    See Church & Gale (1995), *Inverse Document Frequency (IDF): A Measure
    of Deviation from Poisson*.

    ``RIDF = IDF_observed - IDF_expected``

    where:
    - ``IDF = log(N / df)`` (natural log, matching Java ``Math.log``)
    - ``attf = ttf / N``
    - ``pi = exp(-attf)``
    - ``IDF_expected = -log(1 - pi) / log(2)`` (log base 2)
    """

    @property
    def description(self) -> str:
        return "Residual IDF: deviation from Poisson model"

    def score(
        self,
        candidates: list[Candidate],
        term_freq: TermFrequency,
        **kwargs: Any,
    ) -> TermExtractionResult:
        total_docs = float(term_freq.total_docs)
        result = TermExtractionResult()

        for candidate in candidates:
            nf = candidate.normalized_form
            ttf = term_freq.get_ttf(nf)
            df = term_freq.get_df(nf)

            if df == 0 or total_docs == 0:
                result.add(
                    Term(
                        string=candidate.normalized_form,
                        score=0.0,
                        frequency=ttf,
                        surface_forms=set(candidate.surface_forms),
                    )
                )
                continue

            attf = ttf / total_docs
            pi = math.exp(-attf)

            # Expected IDF (Poisson-based), log base 2
            if pi >= 1.0:
                eidf = 0.0
            else:
                eidf = -(math.log(1 - pi) / math.log(2))

            # Observed IDF (natural log, matching Java)
            idf = math.log(total_docs / df)

            ridf = idf - eidf

            term = Term(
                string=candidate.normalized_form,
                score=ridf,
                frequency=ttf,
                surface_forms=set(candidate.surface_forms),
            )
            result.add(term)

        return result.sort()
